src/audiobuffer/audiobuffer.cpp

```diff
@@ -0,0 +1,252 @@
+			bytesWritten = mBuffer.addData((int16_t *)aData, aDataLen / mPCMformat.bytesPerSample);
+			break;
+		case 3:
+			bytesWritten = mBuffer.addData((int32_t *)aData, aDataLen / mPCMformat.bytesPerSample);
+			break;
+		}
+
+		mSampleCount += bytesWritten / mPCMformat.bytesPerSample;
+
+		// If a handle reaches the end and data is not ended, we have to wait for it has enough data
+		// to reach [TIME_FOR_BUFFERING] and restart playing it.
+		// time currBufferTime = 1.0f * mBuffer.getFloatsBufferSize() / mPCMformat.channels / mPCMformat.sampleRate;
+		time currBufferTime = getLength();
+		for (int i = 0; i < mParent->handle.size(); i++)
+		{
+			double pos = mThePlayer->getPosition(mParent->handle[i].handle);
+			// This handle needs to wait for [TIME_FOR_BUFFERING]
+			if (pos >= currBufferTime && !mThePlayer->getPause(mParent->handle[i].handle))
+			{
+				mParent->handle[i].bufferingTime = currBufferTime;
+				mThePlayer->setPause(mParent->handle[i].handle, true);
+				if (mOnBufferingCallback != nullptr)
+				{
+#ifdef __EMSCRIPTEN__
+						// Call the Dart callback stored on globalThis, if it exists.
+						EM_ASM({
+							// Compose the function name for this soundHash
+							var functionName = "dartOnBufferingCallback_" + $3;
+							if (typeof window[functionName] === "function") {
+								var buffering = $0 == 1 ? true : false;
+								window[functionName](buffering, $1, $2); // Call it
+							} else {
+								console.log("EM_ASM 'dartOnBufferingCallback_$hash' not found.");
+							}
+						}, true, mParent->handle[i].handle, currBufferTime, mParent->soundHash);
+#else
+						mOnBufferingCallback(true, mParent->handle[i].handle, currBufferTime);
+#endif
+				}
+			}
+			if (currBufferTime - mParent->handle[i].bufferingTime >= mBufferingTimeNeeds && 
+				mThePlayer->getPause(mParent->handle[i].handle))
+			{
+				mThePlayer->setPause(mParent->handle[i].handle, false);
+				mParent->handle[i].bufferingTime = MAX_DOUBLE;
+				if (mOnBufferingCallback != nullptr)
+				{
+#ifdef __EMSCRIPTEN__
+						// Call the Dart callback stored on globalThis, if it exists.
+						EM_ASM({
+							// Compose the function name for this soundHash
+							var functionName = "dartOnBufferingCallback_" + $3;
+							if (typeof window[functionName] === "function") {
+								var buffering = $0 == 1 ? true : false;
+								window[functionName](buffering, $1, $2); // Call it
+							} else {
+								console.log("EM_ASM 'dartOnBufferingCallback_$hash' not found.");
+							}
+						}, false, mParent->handle[i].handle, currBufferTime, mParent->soundHash);
+#else
+						mOnBufferingCallback(false, mParent->handle[i].handle, currBufferTime);
+#endif
+				}
+			}
+		}
+
+
+		// data has been added to the buffer, but not all because reached its full capacity.
+		// So mark this stream as ended and no more data can be added.
+		if (bytesWritten < aDataLen / mPCMformat.bytesPerSample)
+		{
+			dataIsEnded = true;
+			return pcmBufferFullOrStreamEnded;
```

Comment:
I'd prefer having two separate errors: one for full buffer, second for stream ended. This makes it much easier for the developer to debug problems, as they don't need to investigate 2 options instead of one.

Reply:
Done.